refactor: share `FlatOptions` in whole build session

IWANABETHATGUY opened the pull request:
### TL;DR

Refactored `ScannerFlatOptions` to `FlatOptions` and improved its propagation throughout the codebase.

### What changed?

- Renamed `ScannerFlatOptions` to `FlatOptions` to better reflect its purpose
- Added documentation to clarify that `FlatOptions` is used to avoid passing `&SharedNormalizedBundlerOptions` everywhere and to make accessing frequently used options faster
- Modified the `AstScanner` to use `flat_options` instead of `flags`
- Updated the module loader to create and propagate `flat_options` to various tasks
- Ensured `flat_options` is passed through the scan stage output to the link stage
- Updated all references to use the renamed struct

### How to test?

- Run the existing test suite to ensure all functionality works as expected
- Verify that all modules using the options still behave correctly
- Check that performance is maintained or improved with the refactored approach

### Why make this change?

This change improves code organization and readability by:
1. Using a more appropriate name for the options struct
2. Ensuring consistent propagation of options throughout the codebase
3. Making the purpose of the struct clearer with added documentation
4. Reducing redundant option resolution by passing the flat options structure through the build pipeline

The refactoring should make the codebase more maintainable while potentially improving performance by avoiding repeated function calls to access options.

IWANABETHATGUY commented:
* **#6207** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6207?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6211** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6211?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6211?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6211).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

IWANABETHATGUY commented on crates/rolldown/src/stages/link_stage/mod.rs:
```diff
@@ -80,6 +81,8 @@ pub struct LinkStage<'a> {
   pub entry_point_to_reference_ids: FxHashMap<EntryPoint, Vec<ArcStr>>,
   pub global_constant_symbol_map: FxHashMap<SymbolRef, ConstExportMeta>,
+  #[expect(dead_code)]
+  pub flat_options: FlatOptions,
```
Used for `cross_module_optimization`

hyf0 reviewed (approved):
What does `flat` mean here?

IWANABETHATGUY commented:
> What does `flat` mean here?

Flatten all frequently accessed option fields in `NormalizedOptions` into a `BitFlag`, before it may be nested, so I use `flat` here

IWANABETHATGUY committed:
refactor: share `FlatOptions` in whole build session (#6211)

### TL;DR

Refactored `ScannerFlatOptions` to `FlatOptions` and improved its propagation throughout the codebase.

### What changed?

- Renamed `ScannerFlatOptions` to `FlatOptions` to better reflect its purpose
- Added documentation to clarify that `FlatOptions` is used to avoid passing `&SharedNormalizedBundlerOptions` everywhere and to make accessing frequently used options faster
- Modified the `AstScanner` to use `flat_options` instead of `flags`
- Updated the module loader to create and propagate `flat_options` to various tasks
- Ensured `flat_options` is passed through the scan stage output to the link stage
- Updated all references to use the renamed struct

### How to test?

- Run the existing test suite to ensure all functionality works as expected
- Verify that all modules using the options still behave correctly
- Check that performance is maintained or improved with the refactored approach

### Why make this change?

This change improves code organization and readability by:
1. Using a more appropriate name for the options struct
2. Ensuring consistent propagation of options throughout the codebase
3. Making the purpose of the struct clearer with added documentation
4. Reducing redundant option resolution by passing the flat options structure through the build pipeline

The refactoring should make the codebase more maintainable while potentially improving performance by avoiding repeated function calls to access options.
